crates/biome_js_analyze/src/lint/nursery/no_irregular_whitespace.rs

```diff
@@ -0,0 +1,110 @@
+use biome_analyze::{context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_syntax::AnyJsStatement;
+use biome_rowan::{AstNode, TextRange, TextSize};
+
+const IRREGULAR_WHITESPACES: &[char; 22] = &[
+    '\u{c}', '\u{b}', '\u{85}', '\u{feff}', '\u{a0}', '\u{1680}', '\u{180e}', '\u{2000}',
+    '\u{2001}', '\u{2002}', '\u{2003}', '\u{2004}', '\u{2005}', '\u{2006}', '\u{2007}', '\u{2008}',
+    '\u{2009}', '\u{200a}', '\u{200b}', '\u{202f}', '\u{205f}', '\u{3000}',
+];
+
+declare_rule! {
+    /// Disallows the use of irregular whitespace characters.
+    ///
+    /// Invalid or irregular whitespace causes issues with ECMAScript 5 parsers and also makes code harder to debug.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// constcount=1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = 'thing';
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const count = 1;
+    /// ```
+    ///
+    /// ```js
+    /// const foo = '';
+    /// ```
+    ///
+    pub NoIrregularWhitespace {
+        version: "next",
+        name: "noIrregularWhitespace",
+        language: "js",
+        recommended: false,
+    }
+}
+
+impl Rule for NoIrregularWhitespace {
+    type Query = Ast<AnyJsStatement>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        get_irregular_whitespace(node)
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Irregular whitespaces found."
+                },
+            )
+            .note(markup! {
+                "Replace the irregular whitespaces with normal whitespaces or tabs."
+            }),
+        )
+    }
+}
+
+fn get_irregular_whitespace(node: &AnyJsStatement) -> Option<()> {
+    let syntax = node.syntax();
+    let node_text = syntax.text_trimmed();
+    let range_start: u32 = node.range().start().into();
```

Comment:
I believe we can evaluate a different approach. As you might notice [in this example](https://biomejs.dev/playground/?code=YwBvAG4AcwB0AAsAZgBvAG8ACwA9AAsAJwB0AGgAaQBuAGcAJwA7AA%3D%3D), the irregular whitespaces are inside **trivia**. 

This information is already present inside the nodes, inside the trivia attached to those nodes. You can retrieve trivia from **tokens**. You have multiple ways to retrieve the tokens: you directly query all tokens via `Ast<JsSyntaxToken>`, use some methods from `syntax` (there are many depending on the situation).

Reply:
oh, I didn't know we can query all tokens like that, that definitely makes more sense, I will look into that :)